server/src/main/java/org/opentosca/toscana/plugins/cloudformation/CloudFormationModule.java

```diff
@@ -0,0 +1,136 @@
+package org.opentosca.toscana.plugins.cloudformation;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.opentosca.toscana.core.plugin.PluginFileAccess;
+
+import com.scaleset.cfbuilder.core.Fn;
+import com.scaleset.cfbuilder.core.Module;
+import com.scaleset.cfbuilder.core.Resource;
+import com.scaleset.cfbuilder.core.Template;
+import com.scaleset.cfbuilder.ec2.Instance;
+import com.scaleset.cfbuilder.ec2.UserData;
+import com.scaleset.cfbuilder.ec2.metadata.CFNInit;
+
+public class CloudFormationModule extends Module {
+
+    public final static String CONFIG_SETS = "InstallAndConfigure";
+    public final static String CONFIG_INSTALL = "Install";
+    public final static String CONFIG_CONFIGURE = "Configure";
+    public final static String SECURITY_GROUP = "SecurityGroup";
+    // KeyName is an default input value
+    private static final String KEYNAME_DESCRIPTION = "Name of an existing EC2 KeyPair to enable SSH access to the instances";
+    private static final String KEYNAME_TYPE = "AWS::EC2::KeyPair::KeyName";
+    private static final String KEYNAME_CONSTRAINT_DESCRIPTION = "must be the name of an existing EC2 KeyPair.";
+    private static final String KEYNAME = "KeyName";
+    
+    private Object keyNameVar;
+
+    private Map<String, CFNInit> cfnInitMap;
+
+    private PluginFileAccess fileAccess;
+
+    /**
+     Create a Module which uses the cloudformation-builder to build an AWS CloudFormation template
+
+     @param fileAccess fileAccess to append the content of files to the template
+     */
+    public CloudFormationModule(PluginFileAccess fileAccess) {
+        this.id("").template(new Template());
+        strParam(KEYNAME).type(KEYNAME_TYPE).description(KEYNAME_DESCRIPTION).constraintDescription(KEYNAME_CONSTRAINT_DESCRIPTION);
+        keyNameVar = template.ref(KEYNAME);
+        cfnInitMap = new HashMap<>();
+        this.fileAccess = fileAccess;
+    }
+
+    /**
+     Put a CFNInit into a map which will be added to the resource at build time
+     @param resource resource to add CFNInit to
+     @param init CNFInit to add
+     */
+    public void putCFNInit(String resource, CFNInit init) {
+        cfnInitMap.put(resource, init);
+    }
+
+    /**
+     Get the CFNInit which belongs to a specific resource
+
+     @param resource String id of a resource
+     */
+    public CFNInit getCFNInit(String resource) {
+        return this.cfnInitMap.get(resource);
+    }
+
+    /**
+     Get a ref to the KeyName of this template
+     */
+    public Object getKeyNameVar() {
+        return this.keyNameVar;
+    }
+
+    private Fn getUserDataFn(String resource, String configsets) {
+        return Fn.fnDelimiter("Join", "",
```

Comment:
This looks bad. is it possible to put his into a constant?

if not just leave it as is

Reply:
we'll look into this

Reply:
Instead of putting the whole thing into a constant, I split the constant part and the part that needs the method parameters up in order to make the method more clear.